pkg/test/ginkgo/queue.go

```diff
@@ -12,141 +13,155 @@ import (
+				return
+			}
+			// if the context is finished, simply return
+			if ctx.Err() != nil {
+				return
+			}
+			testSuiteRunner.RunOneTest(ctx, test)
 		}
-		defer q.done(test)
-		fn(ctx, test)
-		return true
 	}
 }
 
-func (q *parallelByFileTestQueue) Execute(ctx context.Context, tests []*testCase, parallelism int, testFn TestFunc) {
-	defer q.Close()
+// tests are currently being mutated during the run process.
+func (q *parallelByFileTestQueue) Execute(ctx context.Context, tests []*testCase, parallelism int, testOutput testOutputConfig, maybeAbortOnFailureFn testAbortFunc) {
+	testSuiteProgress := newTestSuiteProgress(len(tests))
+	testSuiteRunner := &testSuiteRunnerImpl{
+		commandContext:        q.commandContext,
+		testOutput:            testOutput,
+		testSuiteProgress:     testSuiteProgress,
+		maybeAbortOnFailureFn: maybeAbortOnFailureFn,
+	}
+
+	execute(ctx, testSuiteRunner, tests, parallelism)
+}
 
+// execute is a convenience for unit testing
+func execute(ctx context.Context, testSuiteRunner testSuiteRunner, tests []*testCase, parallelism int) {
 	if ctx.Err() != nil {
 		return
 	}
 
-	serial, parallel := splitTests(tests, func(t *testCase) bool { return strings.Contains(t.name, "[Serial]") })
+	serial, parallel := splitTests(tests, isSerialTest)
 
-	r := ring.New(len(parallel))
-	for _, test := range parallel {
-		r.Value = test
-		r = r.Next()
-	}
-	q.queue = r
+	remainingParallelTests := make(chan *testCase, 100)
+	go queueAllTests(remainingParallelTests, parallel)
 
 	var wg sync.WaitGroup
-	wg.Add(parallelism)
 	for i := 0; i < parallelism; i++ {
-		go func(i int) {
+		wg.Add(1)
+		go func(ctx context.Context) {
 			defer wg.Done()
-			for q.Take(ctx, func(ctx context.Context, test *testCase) {
-				testFn(ctx, test)
-			}) {
-				if ctx.Err() != nil {
-					return
-				}
-			}
-		}(i)
+			runTestsUntilChannelEmpty(ctx, remainingParallelTests, testSuiteRunner)
+		}(ctx)
 	}
 	wg.Wait()
 
 	for _, test := range serial {
 		if ctx.Err() != nil {
 			return
 		}
-		testFn(ctx, test)
+		testSuiteRunner.RunOneTest(ctx, test)
 	}
 }
 
-func setTestExclusion(tests []*testCase, fn func(suitePath string, t *testCase) bool) {
-	for _, test := range tests {
-		summary := test.spec.Summary("")
-		var suitePath string
-		for _, loc := range summary.ComponentCodeLocations {
-			if len(loc.FileName) > 0 {
-				if !strings.HasSuffix(loc.FileName, "/k8s.io/kubernetes/test/e2e/framework/framework.go") {
-					suitePath = loc.FileName
-				}
+func isSerialTest(test *testCase) bool {
+	if strings.Contains(test.name, "[Serial]") {
+		return true
+	}
+
+	// Original Clayton comment
+	// This ensures that tests in the identified paths do not run in parallel, because
+	// the test suite reuses shared resources without considering whether another test
+	// could be running at the same time. While these are technically [Serial], ginkgo
+	// parallel mode provides this guarantee. Doing this for all suites would be too
```

Comment:
I don't get this comment, it's very confusing. 

```
While these are technically [Serial], ginkgo parallel mode provides this guarantee.
```

What guarantee?

Reply:
> I don't get this comment, it's very confusing.
>While these are technically [Serial], ginkgo parallel mode provides this guarantee.
What guarantee?

TBH, I don't actually know.  It came from clayton and I preserved it.

Reply:
Continuing this thread, according to maciej, we don't even need this block because it has been broken for four releases.